### PR TITLE
Refactor tools and tools bucket

### PR DESCRIPTION
**Pull request by metonymic-smokey**
This PR is about refactoring `tools.go` and `tools_bucket.go` based on #2248.

* [ ] I added CHANGELOG entry for this change.
* [x] Change is not relevant to the end user.

## Changes
Added config struct for flags in `tools.go`.
Doing the same for `tools_bucket.go`

## Verification
Building the thanos CLI works as expected.

### CODE REVIEW

**Comment by metonymic-smokey** [comment]
@yeya24 @squat could you review this WIP PR?
Working on doing the same for `tools_bucket.go`
Thanks!

**Comment by metonymic-smokey** [comment]
Current status: Moved all the flags in `tools_bucket.go` into one config struct. Each subcommand has a separate function to register flags. 
Since there are too many flags in the struct now, would it be better to move them into separate structs(one for each subcommand)?

File: cmd/thanos/tools.go [hunk 1]
```diff
@@ -18,21 +18,30 @@ import (
...
 	cmd := app.Command("rules-check", "Check if the rule files are valid or not.")
-	ruleFiles := cmd.Flag("rules", "The rule files glob to check (repeated).").Required().Strings()
-
+	tc := &toolsConfig{}
+	tc.registerFlag(cmd)
 	cmd.Setup(func(g *run.Group, logger log.Logger, reg *prometheus.Registry, _ opentracing.Tracer, _ <-chan struct{}, _ bool) error {
 		// Dummy actor to immediately kill the group after the run function returns.
 		g.Add(func() error { return nil }, func(error) {})
-		return checkRulesFiles(logger, ruleFiles)
+		return checkRulesFiles(logger, &(tc.rulesFiles))
```
Comment (GiedriusS) [inline cmd/thanos/tools.go:44]: Extra parentheses aren't needed here (:

File: cmd/thanos/tools_bucket.go [hunk 2]
```diff
@@ -142,7 +278,7 @@ func registerBucketVerify(app extkingpin.AppClause, objStoreConfig *extflag.Path
 		// Dummy actor to immediately kill the group after the run function returns.
 		g.Add(func() error { return nil }, func(error) {})
 
-		r, err := issuesVerifiersRegistry.SubstractByIDs(*issuesToVerify, *repair)
+		r, err := issuesVerifiersRegistry.SubstractByIDs(tbc.issuesToVerify, (tbc.repair))
```
Comment (GiedriusS) [inline cmd/thanos/tools_bucket.go:281]: Ditto

**Review by GiedriusS — COMMENTED** [review]
> Current status: Moved all the flags in `tools_bucket.go` into one config struct. Each subcommand has a separate function to register flags.
> Since there are already too many flags in the struct now, would it be better to move them into separate structs(one for each subcommand)?

I wanted to say the exact same thing after reading your code :smile: Yep, I think it would be clearer because now different members have no connection between them

**Comment by GiedriusS** [comment]
Just those two small comments and I think we can merge this. Thank you :+1:

File: cmd/thanos/tools.go (hunk 1 above)
Comment (metonymic-smokey) [reply cmd/thanos/tools.go]: resolved in latest commit.

File: cmd/thanos/tools_bucket.go (hunk 2 above)
Comment (metonymic-smokey) [reply cmd/thanos/tools_bucket.go]: resolved in latest commit.

**Comment by metonymic-smokey** [comment]
Thanks for the review @GiedriusS!
My latest commit should make the code clearer. Please let me know if there is something more I can do.

**Comment by metonymic-smokey** [comment]
I'm not sure why my Thanos E2E tests are failing with the error 'Flaky test'.

File: cmd/thanos/tools.go [hunk 3]
```diff
@@ -18,21 +18,30 @@ import (
 	"github.com/thanos-io/thanos/pkg/rules"
 )
 
+type toolsConfig struct {
```
Comment (yeya24) [inline cmd/thanos/tools.go:21]: It is better to be more specific here. What about `checkRulesConfig`?

**Review by yeya24 — DISMISSED** [review]
Awesome work and just one small nit! Thank you @metonymic-smokey for the high quality pr!

**Comment by yeya24** [comment]
> I'm not sure why my Thanos E2E tests are failing with the error 'Flaky test'.

It is not related to your changes so it is fine.

File: cmd/thanos/tools.go (hunk 3 above)
Comment (metonymic-smokey) [reply cmd/thanos/tools.go]: done in my latest commit!

**Review by yeya24 — APPROVED** [review]
LGTM

**Pull request merged**